File: backend/app/models/player_season.rb
class PlayerSeason < ApplicationRecord
  acts_as_paranoid

  enum growth_type: { super_early: 0, early: 1, normal: 2, late: 3, super_late: 4 }
  enum current_growth_type: { pre_growth: 0, growth_phase: 1, prime: 2, decline_phase: 3 }

  # バリデーション
  validates :age, presence: true, inclusion: 16..99, numericality: { only_integer: true }
  validates :year, presence: true, inclusion: 2000..9999
  validates :number, length: { maximum: 3 }, allow_nil: true
  validates :plate_appearances, length: { maximum: 80 }
  validates :memo, length: { maximum: 80 }
  validates :growth_type, :current_growth_type, presence: true

  # アソシエーション
  belongs_to :player
  has_many :pitcher_seasons, dependent: :destroy
  has_many :batter_seasons, dependent: :destroy
  has_many :pitcher_abilities, dependent: :destroy
  has_many :batter_abilities, dependent: :destroy
  has_many :breaking_balls, dependent: :destroy
  has_many :special_abilities, dependent: :destroy

    # 削除
    def delete_item
      update(deleted_at: Time.current)
    end
  
    # 詳細取得
    def self.get_item(id)
      find_by(id: id, deleted_at: nil)
    end

    # 投手の選手記録一覧取得
    def self.get_pitcher_all_item(id)
      find_by(id: id, deleted_at: nil)
    end
end

File: frontend/src/models/player_season.ts
export type PlayerSeason = {
    id?: number;
    year: number | null;
    age: number | null;
    number: string | null;
    growth_type: string | null;
    current_growth_type: string | null;
    is_starter: boolean;
    is_relief: boolean;
    is_closer: boolean;
    is_catcher: boolean;
    is_first: boolean;
    is_second: boolean;
    is_third: boolean;
    is_short: boolean;
    is_outfielder: boolean;
    plate_appearances: string;
    memo?: string;
};

export const createPlayerSeason = (overrides?: Partial<PlayerSeason>): PlayerSeason => ({
    year: null,
    age: null,
    number: null,
    growth_type: 'normal',
    current_growth_type: 'normal',
    is_starter: false,
    is_relief: false,
    is_closer: false,
    is_catcher: false,
    is_first: false,
    is_second: false,
    is_third: false,
    is_short: false,
    is_outfielder: false,
    plate_appearances: '',
    memo: '',
    ...overrides,
});

// 送信フォーム用
export type PlayerSeasonForm = {
  year: string;
  age: string;
  number: string;
  growth_type: string;
  current_growth_type: string;
  is_starter: string;
  is_relief: string;
  is_closer: string;
  is_catcher: string;
  is_first: string;
  is_second: string;
  is_third: string;
  is_short: string;
  is_outfielder: string;
  plate_appearances: string;
  memo: string;
};

export const buildPlayerSeasonPayload = (
  form: PlayerSeasonForm
): PlayerSeason => ({
  year: form.year ? Number(form.year) : 0,
  age: form.age ? Number(form.age) : 0,
  number: form.number || '',
  growth_type: form.growth_type || 'normal',
  current_growth_type: form.current_growth_type || 'decline_phase',
  is_starter: form.is_starter === "true",
  is_relief: form.is_relief === "true",
  is_closer: form.is_closer === "true",
  is_catcher: form.is_catcher === "true",
  is_first: form.is_first === "true",
  is_second: form.is_second === "true",
  is_third: form.is_third === "true",
  is_short: form.is_short === "true",
  is_outfielder: form.is_outfielder === "true",
  plate_appearances: form.plate_appearances || '',
  memo: form.memo || ''
});